src/Playwright/Core/Locator.cs

```diff
@@ -338,4 +342,249 @@ private async Task WithElementAsync(Func<IElementHandle, float?, Task> callback,
+
+    internal static string GetByAttributeTextSelector(string attrName, string text, bool? exact)
+    {
+        var exactFlag = (exact == true) ? "s" : "i";
+        return $"internal:attr=[{attrName}={EscapeForAttributeSelector(text)}{exactFlag}]";
+    }
+
+    internal static string GetByAttributeTextSelector(string attrName, Regex text, bool? exact)
+        => $"internal:attr=[{attrName}={EscapeForTextSelector(text, exact)}]";
+
+    internal static string GetByLabelSelector(string text, bool? exact)
+        => "internal:label=" + EscapeForTextSelector(text, exact);
+
+    internal static string GetByLabelSelector(Regex text, bool? exact)
+        => "internal:label=" + EscapeForTextSelector(text, exact);
+
+    internal static string GetByAltTextSelector(string text, bool? exact)
+        => GetByAttributeTextSelector("alt", text, exact);
+
+    internal static string GetByAltTextSelector(Regex text, bool? exact)
+        => GetByAttributeTextSelector("alt", text, exact);
+
+    internal static string GetByTitleSelector(string text, bool? exact)
+        => GetByAttributeTextSelector("title", text, exact);
+
+    internal static string GetByTitleSelector(Regex text, bool? exact)
+        => GetByAttributeTextSelector("title", text, exact);
+
+    internal static string GetByPlaceholderSelector(string text, bool? exact)
+        => GetByAttributeTextSelector("placeholder", text, exact);
+
+    internal static string GetByPlaceholderSelector(Regex text, bool? exact)
+        => GetByAttributeTextSelector("placeholder", text, exact);
+
+    internal static string GetByTextSelector(string text, bool? exact)
+        => $"text={EscapeForTextSelector(text, exact)}";
+
+    internal static string GetByTextSelector(Regex text, bool? exact)
+        => $"text={EscapeForTextSelector(text, exact)}";
+
+    internal static string GetByRoleSelector(string role, ByRoleOptions options)
+    {
+        List<List<string>> props = new();
+        if (options.Checked != null)
+        {
+            props.Add(new List<string> { "checked", options.Checked.ToString() });
+        }
+        if (options.Disabled != null)
+        {
+            props.Add(new List<string> { "disabled", options.Disabled.ToString() });
+        }
+        if (options.Selected != null)
+        {
+            props.Add(new List<string> { "selected", options.Selected.ToString() });
+        }
+        if (options.Expanded != null)
+        {
+            props.Add(new List<string> { "expanded", options.Expanded.ToString() });
+        }
+        if (options.IncludeHidden != null)
+        {
+            props.Add(new List<string> { "include-hidden", options.IncludeHidden.ToString() });
+        }
+        if (options.Level != null)
+        {
+            props.Add(new List<string> { "level", options.Level.ToString() });
+        }
+        if (options.NameString != null)
+        {
+            props.Add(new List<string> { "name", options.NameString });
```

Comment:
You need to escape the string and regex, use EscapeForAttributeSelector here

Reply:
following up with role stuff.